File: external/dmp/matlab/functionapproximator/ModelParametersLWPR.m
classdef ModelParametersLWPR < Parameterizable
  properties(SetAccess='private')
    ID_
    model_
  end

  methods
    function obj = ModelParametersLWPR(ID, model)
      obj.ID_    = ID;
      obj.model_ = model;
    end

    function selectable_labels = getSelectableModelParameters(obj)
      selectable_labels = {};
      error('Not implemented yet...') %#ok<WNTAG>
    end
    function size = getModelParameterValuesSize(obj)
      size = 0;
      error('Not implemented yet...') %#ok<WNTAG>
    end
    function obj = setSelectedModelParameters(obj,selected_values_labels)
      error('Not implemented yet...') %#ok<WNTAG>
    end
    function values = getModelParameterValues(obj)
      values = [];
      error('Not implemented yet...') %#ok<WNTAG>
    end
    function obj = setModelParameterValues(obj,values)
      error('Not implemented yet...') %#ok<WNTAG>
    end
    function model_parameters_clone = clone(obj)
      model_parameters_clone = 0;
      error('Not implemented yet...') %#ok<WNTAG>
    end
  end  
end


% ModelParametersLWPR(LWPR_Object* lwpr_object)
% :
%   obj.lwpr_object_ = lwpr_object;
% {
%   
%   % Initialize the value_types_ vector
%   n_centers = 0;
%   n_slopes = 0;
%   n_offsets = 0;
% 	for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
% 	{
% 		for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
% 		{
%       n_centers += lwpr_object_->nIn(); 
%       n_slopes  += lwpr_object_->model.sub[iDim].rf[iRF]->nReg;
%       n_offsets += 1;
% end
% end
%   n_widths = n_centers;
%   
%   n_values = n_centers + n_widths + n_slopes + n_offsets;
%   value_types_ = VectorXi(n_values);
%   value_types_ << Constant(n_centers, LWPR_CENTERS),
%                   Constant( n_widths, LWPR_WIDTHS ),
%                   Constant( n_slopes, LWPR_SLOPES ),
%                   Constant(n_offsets, LWPR_OFFSETS );   
% end
%   
% ModelParametersLWPR* clone()  {
%   LWPR_Object* lwpr_object_clone =  LWPR_Object(*lwpr_object_);
%   return  ModelParametersLWPR(lwpr_object_clone);
% end
% 
% ostream& toStream(ostream& output)  {
%   output << "ModelParametersLWPR[todo]";
%   return output;
% end
% 
% istream& fromStream(istream& input) {
%   cerr << _FILE___ << ":" << _LINE___ << ":Not implemented yet!" << endl;
%   return input;
% end
% 
%  getValues(VectorXd& values)  {
%   % First determine the size of the values vector to be returned and resize it.
%   values.resize(length(value_types_));
% 
%   ii=0;
%   
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->nIn(); j++)
%        values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->c[j];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->nIn(); j++)
%         values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->D[j*lwpr_object_->model.nInStore+j];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->model.sub[iDim].rf[iRF]->nReg; j++)
%         values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->beta[j];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->beta0;
% end
% 
% 
%  setValues( VectorXd& values) {
%   if (length(value_types_) != length(values))
%   {
%     cerr << _FILE___ << ":" << _LINE___ << ": values is of wrong size." << endl;
%     return;
% end
% 
%   ii=0;
%   
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->nIn(); j++)
%        lwpr_object_->model.sub[iDim].rf[iRF]->c[j] = values[ii++];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->nIn(); j++)
%         lwpr_object_->model.sub[iDim].rf[iRF]->D[j*lwpr_object_->model.nInStore+j] = values[ii++];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       for (j = 0; j < lwpr_object_->model.sub[iDim].rf[iRF]->nReg; j++)
%         lwpr_object_->model.sub[iDim].rf[iRF]->beta[j] = values[ii++];
%       
%   for (iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
%     for (iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
%       lwpr_object_->model.sub[iDim].rf[iRF]->beta0 = values[ii++];
% end
% 
%  getValueTypes(VectorXi& value_types) 
% { 
%   value_types =  value_types_;
% end
% 
% /*
% %  save( char* file)
% % {
% % 	lwpr_object_->writeBinary(file);
% % }
% */

File: external/dmp/cpp/src/functionapproximators/ModelParametersLWPR.cpp
/**
 * @file   ModelParametersLWPR.h
 * @brief  ModelParametersLWPR class source file.
 * @author Freek Stulp, Thibaut Munzer
 */
 
#include "functionapproximators/ModelParametersLWPR.h"

#include "utilities/ParseJSON.hpp"
#include "utilities/EigenJSON.hpp"

#include "lwpr.hh"
#include <boost/regex.hpp>


#include <iostream>
#include <fstream>

using namespace Eigen;
using namespace std;

ModelParametersLWPR::ModelParametersLWPR(LWPR_Object* lwpr_object)
:
  lwpr_object_(lwpr_object)
{
  countLengths();
}

void ModelParametersLWPR::countLengths(void)
{
 
  // Determine the lengths of different vectors
  n_centers_ = 0;
  n_slopes_ = 0;
  n_offsets_ = 0;
	for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
	{
		for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
		{
      n_centers_ += lwpr_object_->nIn(); 
      n_slopes_  += lwpr_object_->model.sub[iDim].rf[iRF]->nReg;
      n_offsets_ += 1;
		}
	}
  n_widths_ = n_centers_;
  
}
  
ModelParametersLWPR::~ModelParametersLWPR(void)
{
  delete lwpr_object_;
}

ModelParameters* ModelParametersLWPR::clone(void) const {
  LWPR_Object* lwpr_object_clone = new LWPR_Object(*lwpr_object_);
  return new ModelParametersLWPR(lwpr_object_clone);
}

ostream& ModelParametersLWPR::serialize(ostream& output) const {
  output << "{ \"ModelParametersLWPR\": { \"lwpr_object_xml_string\": \"";

  // Write file
  string filename("/tmp/lpwrfile_serialize.xml");
  lwpr_object_->writeXML(filename.c_str());
  
  // Read file into string stream (through file stream)
  ifstream t(filename);
  stringstream strstream;
  strstream << t.rdbuf(); 

  // Get the string and remove newlines  
  string str = strstream.str();
  str.erase(std::remove(str.begin(), str.end(), '\n'), str.end());
  
  // Output the XML string to the current output stream
  output << str;
  
  output << " \" } }";

  return output;
};


ModelParametersLWPR* ModelParametersLWPR::deserialize(istream& input_stream)
{
  bool remove_white_space = true;
  string json_string = getJSONString(input_stream,remove_white_space);
  
  if (json_string.empty())
    return NULL;
  
  string lwpr_object_xml_string;
  
  lwpr_object_xml_string = getJSONValue(json_string,"lwpr_object_xml_string");
  
  // We now have the LWPR object as an XML string.
  // Save this XML string to a temporary file, and call the LWPR_Object constructor with 
  // the filename. This constructs the LWPR_Object from the XML.
  
  string filename("/tmp/lpwrfile_deserialize.xml");
  ofstream temp(filename);
  temp << lwpr_object_xml_string;
        
  LWPR_Object* lwpr_object = new LWPR_Object(filename.c_str());
  return new ModelParametersLWPR(lwpr_object);

  return NULL;
}

int ModelParametersLWPR::getExpectedInputDim(void) const
{
  return lwpr_object_->nIn();
}

void ModelParametersLWPR::getSelectableParameters(set<string>& selected_values_labels) const 
{
  selected_values_labels = set<string>();
  selected_values_labels.insert("centers");
  selected_values_labels.insert("widths");
  selected_values_labels.insert("slopes");
  selected_values_labels.insert("offsets");
}




void ModelParametersLWPR::getParameterVectorMask(const std::set<std::string> selected_values_labels, VectorXi& selected_mask) const
{
  selected_mask.resize(getParameterVectorAllSize());
  selected_mask.fill(0);
  
  int offset = 0;
  int size;
  
  // Centers
  size = n_centers_;
  if (selected_values_labels.find("centers")!=selected_values_labels.end())
    selected_mask.segment(offset,size).fill(1);
  offset += size;
  
  // Widths
  size = n_widths_;
  if (selected_values_labels.find("widths")!=selected_values_labels.end())
    selected_mask.segment(offset,size).fill(2);
  offset += size;
  
  // Offsets
  size = n_offsets_;
  if (selected_values_labels.find("offsets")!=selected_values_labels.end())
    selected_mask.segment(offset,size).fill(3);
  offset += size;

  // Slopes
  size = n_slopes_;
  if (selected_values_labels.find("slopes")!=selected_values_labels.end())
    selected_mask.segment(offset,size).fill(4);
  offset += size;

  assert(offset == getParameterVectorAllSize());   
}

void ModelParametersLWPR::getParameterVectorAll(VectorXd& values) const
{
  values.resize(getParameterVectorAllSize());
  int ii=0;
  
  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->nIn(); j++)
       values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->c[j];

  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->nIn(); j++)
        values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->D[j*lwpr_object_->model.nInStore+j];

  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->model.sub[iDim].rf[iRF]->nReg; j++)
        values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->beta[j];
      
  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      values[ii++] = lwpr_object_->model.sub[iDim].rf[iRF]->beta0;
  
  assert(ii == getParameterVectorAllSize()); 
  
};

void ModelParametersLWPR::setParameterVectorAll(const VectorXd& values) {
  if (getParameterVectorAllSize() != values.size())
  {
    cerr << __FILE__ << ":" << __LINE__ << ": values is of wrong size." << endl;
    return;
  }
  
  int ii=0;
  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->nIn(); j++)
      lwpr_object_->model.sub[iDim].rf[iRF]->c[j] = values[ii++];

  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->nIn(); j++)
      lwpr_object_->model.sub[iDim].rf[iRF]->D[j*lwpr_object_->model.nInStore+j] = values[ii++];

  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      for (int j = 0; j < lwpr_object_->model.sub[iDim].rf[iRF]->nReg; j++)
        lwpr_object_->model.sub[iDim].rf[iRF]->beta[j] = values[ii++];

  for (int iDim = 0; iDim < lwpr_object_->model.nOut; iDim++)
    for (int iRF = 0; iRF < lwpr_object_->model.sub[iDim].numRFS; iRF++)
      lwpr_object_->model.sub[iDim].rf[iRF]->beta0 = values[ii++];

  assert(ii == getParameterVectorAllSize());   
};

/*
// void FunctionApproximatorLWPR::save(const char* file)
// {
// 	lwpr_object_->writeBinary(file);
// }
*/
